fri/src/verifier.rs

```diff
@@ -2,45 +2,45 @@ use alloc::vec;
 use alloc::vec::Vec;
 
 use itertools::izip;
-use p3_challenger::{CanObserve, CanSampleBits, FieldChallenger, GrindingChallenger};
+use p3_challenger::{CanObserve, CanSample, CanSampleBits, GrindingChallenger};
 use p3_commit::Mmcs;
 use p3_field::{AbstractField, Field, TwoAdicField};
 use p3_matrix::Dimensions;
-use p3_util::{log2_strict_usize, reverse_bits_len};
+use p3_util::reverse_bits_len;
 
-use crate::{FriConfig, FriProof, InputOpening, QueryProof};
+use crate::{FriConfig, FriProof, QueryProof};
 
 #[derive(Debug)]
-pub enum VerificationError<InputMmcsErr, CommitMmcsErr> {
+pub enum VerificationError<CommitMmcsErr> {
     InvalidProofShape,
-    InputMmcsError(InputMmcsErr),
     CommitPhaseMmcsError(CommitMmcsErr),
     FinalPolyMismatch,
     InvalidPowWitness,
 }
 
 pub type VerificationErrorForFriConfig<FC> = VerificationError<
-    <<FC as FriConfig>::InputMmcs as Mmcs<<FC as FriConfig>::Val>>::Error,
     <<FC as FriConfig>::CommitPhaseMmcs as Mmcs<<FC as FriConfig>::Challenge>>::Error,
 >;
 
 type VerificationResult<FC, T> = Result<T, VerificationErrorForFriConfig<FC>>;
 
-pub(crate) fn verify<FC: FriConfig>(
+#[derive(Debug)]
+pub struct FriChallenges<Challenge> {
+    pub query_indices: Vec<usize>,
+    betas: Vec<Challenge>,
+}
+
+pub fn verify_shape_and_sample_challenges<FC: FriConfig>(
```

Comment:
This is actually kind of annoying: FRI `verify` is now two functions, since you need to get the query indices and openings halfway through verification. I mean, this is basically the whole reason for the old virtual MMCS, but I think it's worth reifying the complexity here for efficiency gains. Plus, FRI is more of a backend thing and will be immediately wrapped by one or two PCS's that take care of it.